### DISCUSSION TITLE
CubeCell and setNodeAddress

### DISCUSSION THREAD

**Discussion by d00m178** [jgromes/RadioLib#837] [Q&A]
hi all.
looking for simple example for CubeCell AB01/AB02 which send/receive data with addresses of device.
I would like to send/receive data and know which exactly device send the data.

as far I understand, this method setNodeAddress shoud help me.
https://github.com/jgromes/RadioLib/blob/master/src/modules/SX126x/SX126x.h#L844

for example - I want to set for one device address "111" (server), for second - "222" (client) and then start to wait for data on the server from client. and I need receive data on server only if the client send it to server address "111".

I saw the SX126x_FSK_Modem.ino but unfortunately it didn't help me much..
Please advice if this possible to do.

**Comment by jgromes** [comment]
The way FSK address on SX126x works is basically a filter. Packets with addresses that match the filter are accepted, other are rejected. Every node can accept to up to two addresses - its own address, and the broadcast address.

If you're not interested in broadcast, then on the "server" you can call `setNodeAddress(111);`. Then, from the client, if you want to send data to the server, you will have to `transmit(data, len, 111);`

EDIT: It should be noted all of this is only applicable to FSK - LoRa packet has no address field, so if you want that functionality, you will have to implement it yourself.

**Comment by d00m178** [comment]
> Packets with addresses that match the filter are accepted, other are rejected.
thank you, it is ok for me.

> It should be noted all of this is only applicable to FSK - LoRa packet has no address field, so if you want that functionality, you will have to implement it yourself.

but I dont understand what did you mean.. 
I dont need LoRaWAN, but I need basic LoRa - I just need send the data from one CubeCell HTCC-AB01/02 to another via onboard LoRa radiomodule.
Will it work?
Ideally - if I can send the data CubeCell and Moteino devices. but I will check this later.

**Reply by jgromes** [reply]
I did not mention LoRaWAN at all.

In order to use the built-in address filtering, you have to use FSK modulation. It will not work for LoRa. I would suggest taking a quick look at the SX126x datasheet.

**Comment by d00m178** [comment]
unfortunately it seems that this FSK modulation is not what Im looking for..
I use Moteino (arduino + RFM95) and use these examples from RadioHead library:

[rf95_reliable_datagram_client.pde](https://www.airspayce.com/mikem/arduino/RadioHead/rf95_reliable_datagram_client_8pde-example.html)
[rf95_reliable_datagram_server.pde](https://www.airspayce.com/mikem/arduino/RadioHead/rf95_reliable_datagram_server_8pde-example.html)

I suppose in this way RFM95 use LoRa and library has address filtering capability..
but seems SX1262 and library from CubeCell and Radiolib library somewhat doesnt have this capability..

**Reply by jgromes** [reply]
> unfortunately it seems that this FSK modulation is not what Im looking for..

How did you arrive at this conclusion?

> I use Moteino (arduino + RFM95) and use these examples from RadioHead library:

The examples you linked use RadioHeads built-in "realiable datagram", which adds another protocol layer over the radio module. The RFM9x (SX127x, really) does not have addressing capability in LoRa modulation, only in FSK, same as SX126x. So the addresses you are using have been added by RadioHead library. RadioLib does not include any such thing, and that's by design.

**Comment by d00m178** [comment]
thank you for explanation) 
this is not my subject so I dont know a lot of things..
as far I understand - SX1262 chip can work in LoRa modulation and FSK modulation which are different from each other..
and as far I know - LoRa is "better" in terms of range, sensitivity etc.. 
I definitely will check the FSK mode with CubeCell with Radiolib, but as least now I know than it will be not possible to communicate between CubeCell in FSK and Moteino with Radiohead.. 
which is not good but fair.. somehow I did not mention this important moment - different libraries ..

**Reply by jgromes** [reply]
> SX1262 chip can work in LoRa modulation and FSK modulation which are different from each other..

That is correct.

> LoRa is "better" in terms of range, sensitivity

There's no such thing as inherently "better" modulation. There are only tradeoffs - with LoRa, you are trading off sensitivity/range/power for data rate. However, trying to push LoRa into data rates higher than a few hundred bps leads to much worse results than FSK. I don't know what your use case is, you should carefuly consider things like the data rate, maximum power and occupied bandwidth before choosing one modulation over another.

>  it will be not possible to communicate between CubeCell in FSK and Moteino with Radiohead..

I doubt RadioHead does not implement FSK. And you can always implement your own protocol, even if it is as simple as "adding one number to the packet to specify the destination address".

**Comment by d00m178** [comment]
the thing is that I already have devices based on RadioHead and I want to switch to CubeCell because it has better low-power usage.
so Im looking for the way to use both type of devices. 
and this is not possible because I would need install new firmware on old devices. I thought that somehow both type of devices would communicate each other but now I see - different libraries will prevent this because their concept of "address" is different..
seem the better way is to use cubecell's with radiolib FSk mode only with the same cubecell's. 
or I need think about LoRaWAN and gateway..

**Reply by jgromes** [reply]
If updating devices that are already in the field is not an option, then to keep compatibility, you would have to implement RadioHead's reliable datagram using RadioLib. I'm not familiar with RadioHead too much, I guess it's just going to add some header containing the protocol information.

**Comment by d00m178** [comment]
I have new question - if I anyway go with update all my devices on RMF95 (Moteino) to Radiolib - what is best methods to use to be able to set addresses for nodes? And use this method for Moteino and Cubecell at the same time in one LoRa network? 
Or to have addresses on nodes I need use only FSK modulation?  
Im still dont want to switch to LoRaWAN because it needs to have the gateway - currently this is not an option for my setup.
And I need to have outdoor LoRa nodes with their addresses - 1, 2, 3, ...., N with some sensors. 
These nodes are "sleep" almost all time and during the daytime they wakeup every 1 minute to check if there is request from another node with special address, for example - address 100. 
if there is such request - node send back the data from sensors. If no request - node start to sleep again for 1 min. 
This way to work allows node based on Moteino to work during 3-4 month being covered under snow.
And now I want to add CubeCell.. 
but this issue with addresses prevent to begin my test new modules..
Please advice..

**Reply by jgromes** [reply]
> if I anyway go with update all my devices on RMF95 (Moteino) to Radiolib

I don't know which device you have exactly, but if it's something like ATmega328P then you might find RadioLib only barely fits on it, it's gotten quite large recently.

> what is best methods to use to be able to set addresses for nodes? And use this method for Moteino and Cubecell at the same time in one LoRa network?

If you want to use LoRa, then as I said previously - addresses are not implemented, you have to do it yourself. So you would need to define the packet structure (e.g. have the first byte as the address), and then check that byte for every packet received.
